packages/gatsby/src/utils/page-data.ts

```diff
@@ -256,10 +256,20 @@ export async function handleStalePageData(): Promise<void> {
   })
 
   const deletionPromises: Array<Promise<void>> = []
-  pageDataFilesFromPreviousBuilds.forEach(pageDataFilePath => {
+  const pagePathsToClear = new Set<string>()
+  for (const pageDataFilePath of pageDataFilesFromPreviousBuilds) {
     if (!expectedPageDataFiles.has(pageDataFilePath)) {
+      const stalePageDataContent = await fs.readJson(pageDataFilePath)
```

Comment:
this adds `fs` operation - but just for page-data files that are marked as stale and scheduled for deletion. This is to grab `page.path`.

Alternatively we could try to derive `page.path` from location of `page-data` file, but this is not deterministic due to all the variations of leading and trailing slashes ... and it also would be more convoluted.

Reply:
A nice trick to get this path! 😄